app/api/threads/update_thread.go

```diff
@@ -0,0 +1,304 @@
+package threads
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/France-ioi/validator"
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/formdata"
+
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+)
+
+// threadUpdateFields represents the fields of a thread in database
+type threadUpdateFields struct {
+	// Optional
+	// enum: waiting_for_participant,waiting_for_trainer,closed
+	Status string `json:"status" validate:"helper_group_id_set_if_non_open_to_open_status,oneof=waiting_for_participant waiting_for_trainer closed"` // nolint
+	// Optional
+	HelperGroupID *int64 `json:"helper_group_id" validate:"helper_group_id_not_set_when_set_or_keep_closed,group_visible_by=:user_id,group_visible_by=participant_id,can_request_help_to_when_own_thread"` // nolint
+	// Optional
+	MessageCount *int `json:"message_count" validate:"omitempty,gte=0,exclude_increment_if_message_count_set"`
+}
+
+// updateThreadRequest is the expected input for thread updating
+// swagger:model threadEditRequest
+type updateThreadRequest struct {
+	threadUpdateFields `json:"thread,squash"` // nolint:staticcheck SA5008: unknown JSON option "squash"
+
+	ItemID        int64
+	ParticipantID int64
+
+	// Used to increment the message count when we are not sure of the exact total message count. Can be negative.
+	// Optional
+	MessageCountIncrement *int `json:"message_count_increment"`
+}
+
+// swagger:operation PUT /items/{item_id}/participant/{participant_id}/thread threads threadUpdate
+// ---
+// summary: Update a thread
+// description: >
+//
+//   Service to update thread information.
+//
+//   If the thread doesn't exist, it is created.
+//
+//   Once a thread has been created, it cannot be deleted or set back to `not_started`.
+//
+//   Validations and restrictions:
+//     * if `status` is given:
+//       - The participant of a thread can always switch the thread from open to any another other status.
+//         He can only switch it from non-open to an open status if he is allowed to request help on this item.
+//       - A user who has `can_watch>=answer` on the item AND `can_watch_members` on the participant: can always switch
+//         a thread to any open status (i.e. he can always open it but not close it)
+//       - A user who `can write` on the thread can switch from an open status to another open status.
+//     * if `status` is already "closed" and not changing status OR if switching to status "closed":
+//       `helper_group_id` must not be given
+//     * if switching to an open status from a non-closed status: `helper_group_id` must be given
```

Comment:
The rule is " if switching to an open status from **a non-open status**: helper_group_id must be given" otherwise, this does not make sense.
The implementation seems to implement the correct rule.

Reply:
fixed